fabtests/functional/rdm_stress.c

```diff
@@ -593,58 +592,270 @@ static int run_child(void)
 
 	for (i = 0; i < ctrl_cnt && !ret; i++) {
 		printf("(%d-%d) rpc op %s\n", myid, id_at_server,
-		       rpc_op_str(ctrl[i].op));
-		ret = ctrl_op[ctrl[i].op](&ctrl[i]);
+		       rpc_op_str(ctrls[i].op));
+		ret = ctrl_op[ctrls[i].op](&ctrls[i]);
 	}
 
 free:
 	ft_free_res();
 	return ret;
 }
 
+static bool get_uint64_val(const char *js, jsmntok_t *t, uint64_t *val)
+{
+	if (t->type != JSMN_PRIMITIVE)
+		return false;
+	return (sscanf(&js[t->start], "%lu", val) == 1);
+}
 
-static struct rpc_ctrl ctrl_array[] = {
-	{op_goodbye},
-	{op_hello},
-	{op_msg_req, 1000},
-	{op_msg_resp},
-	{op_tag_req, 2000},
-	{op_tag_resp},
-	{op_msg_req, 1000000},
-	{op_msg_resp},
-	{op_write_req, 5600000, {12000}},
-	{op_write_resp},
-	{op_read_req, 64000},
-	{op_read_resp},
-	{op_tag_req, 2000000},
-	{op_tag_resp},
-	{op_read_req, 32000},
-	{op_read_resp},
-	{op_write_req, 86000, {6000}},
-	{op_write_resp},
-	{op_read_req, 1000000},
-	{op_read_resp},
-	{op_write_req, 56000, {12000}},
-	{op_write_resp},
-	{op_sleep, 100},
-	{op_exit},
-};
+static bool get_op_enum(const char *js, jsmntok_t *t, uint32_t *op)
+{
+	const char *str;
+	size_t len;
+
+	if (t->type != JSMN_STRING)
+		return false;
+
+	str = &js[t->start];
+	len = t->end - t->start;
+
+	if (!strncmp(str, "msg_req", len)) {
```

Comment:
The problem that @j-xiong pointed out is still relevant.  len is relative to the size of str, not "msg_req".  So if the user input is simply "m" or "msg", this will match, when it should be an error.  This should be:
```
strncmp(str, "msg_req", strlen("msg_req"))
```
As suggested, we can define a macro for this.  I'd put any macro in shared.h

Reply:
I can't assume str has strlen("msg_req") characters for checking.  str is a substring of the JSON payload.  I will use a switch on len and then strncmp on strings of the same length.

Reply:
See ubertest/config.c for a similar set of checks.  We could do:
```
(strlen("msg_req") == len) && !strncmp(str, "msg_req", len)
```

Reply:
ubertest's config file is much more complex than the one for rdm_stress.  For now I can get away with using the switch on len.  All strncmps in config.c, except for one, have the problem pointed out by JianXin.  Fixing those will take some time.

Reply:
The reason for using strncmp vs strcmp is to handle the case where one of the strings is not properly terminated.  If we assume that both are, then strcmp is sufficient.  We know that hard-coded strings are properly terminated.  The question is whether we trust the input.  That's why config.c uses this form:
```
strncmp(str, "hard-coded", strlen("hard-coded"))
``
We're treating str as potentially corrupt, and why we shouldn't even call strlen() against it.

Separate from this, I don't like that the proposed code requires the developer to perform strlen() in their head on the strings as part of updating the code.  Having a macro would simplify the code and provide a single place where we could fix the string comparison issue everywhere.
```
#define FT_STR_CHECK(a, b) !(strncmp(a, b, strlen(b) + 1)

if (FT_STR_CHECK(str, "hard-coded"))
```

Reply:
FT_STR_CHECK does not solve the problem JianXin pointed out.  It fails for the case that number of characters you want to check in str is shorter than "hard-coded".  Use the case str contains "hard" without the terminating null and just so happens "-coded" immediately follows str.  Token for str has length of 4 but you are checking 10 and strncmp returns 0.

Reply:
The +1 to the strlen includes the null termination.  You're using the length of the hard-coded string, not the input string. strncmp("hard", "hard-coded", 4) matches.  strncmp("hard", "hard-coded", 10) does not.

I was pointing out that the comparisons in config.c are incorrect.

Reply:
Since the input token length is already available (as calculated by `t->end - t.>start` not via `strlen()`), it doesn't hurt to enforce an exact match:
```
#define FT_TOKEN_CHECK(token, len, keyword) (len == strlen(keyword) && !strncmp(token, keyword, len))

if (FT_TOKEN_CHECK(str, len, "hard-coded"))
```

Reply:
Set the conversation to be "unresolved" so that comments are shown by default.

Reply:
> Since the input token length is already available (as calculated by `t->end - t.>start` not via `strlen()`), it doesn't hurt to enforce an exact match:
> 
> ```
> #define FT_TOKEN_CHECK(token, len, keyword) (len == strlen(keyword) && !strncmp(token, keyword, len))
> 
> if (FT_TOKEN_CHECK(str, len, "hard-coded"))
> ```

I will add this macro to shared.h and use it.  This will keep the code consistent when config.c is fixed.